src/s1ap/handlers/s1ap_msg_delegator.c

```diff
@@ -219,47 +224,84 @@ parse_nas_pdu(char *msg,  int nas_msg_len, struct nasPDU *nas,
+		nas->elements[index].pduElement.ue_network.len = msg[offset];
 		++offset;
-		memcpy((nas->elements[1].ue_network.capab), msg+offset,
-			nas->elements[1].ue_network.len);
-		offset += nas->elements[1].ue_network.len + 1;
-
+		memcpy(
+		 (nas->elements[index].pduElement.ue_network.capab)
+		   ,msg+offset,
+		   nas->elements[index].pduElement.ue_network.len);
+		offset += 
+		 nas->elements[index].pduElement.ue_network.len + 1;
+		
+		index++;
 		/*ESM msg container*/
 		{
-		short len = msg[offset];
-		nas->elements[5].pti = msg[offset + 2];
-		unsigned char val = *(msg+offset+5);
-		/*ESM message header len is 4: bearer_id_flags(1)+proc_tx_id(1)+msg_id(1)
-		 * +pdn_type(1)*/
-		/*element id 13(1101....) = "esm required" flag*/
-		nas->elements[2].esm_info_tx_required = false;
-		if(13 == (val>>4)) {
-			nas->elements[2].esm_info_tx_required = true;
-			if(val & 1) {
-				nas->elements[2].esm_info_tx_required = true;
+		    short len = msg[offset];
+		    nas->elements[index].msgType = 
+		       NAS_IE_TYPE_ESM_MSG;
+		    nas->elements[index].pduElement.esm_msg.procedure_trans_identity = msg[offset + 2];
+		    unsigned char val = *(msg+offset+5);
+		    /*ESM message header len is 4: bearer_id_flags(1)+proc_tx_id(1)+msg_id(1)
+		     * +pdn_type(1)*/
+		    /*element id 13(1101....) = "esm required" flag*/
+		    nas->elements[index].msgType = 
+		       NAS_IE_TYPE_ESM_MSG;
+		    nas->elements[index].pduElement.esm_msg.esm_info_tx_required = false;
+		    if(13 == (val>>4)) 
+		    {
+			if(val & 1) 
+			{
+			    nas->elements[index].pduElement.esm_msg.esm_info_tx_required = true;
 			}
+		    }
+		    offset += len;
 		}
-		offset += len;
-		}
-
-		/*DRX parameter*/
-		offset += 4;
 
-		/*MS network capability*/
-		nas->elements[4].ms_network.element_id = msg[offset];
-		++offset;
-		nas->elements[4].ms_network.len = msg[offset];
-		++offset;
-		memcpy((nas->elements[4].ms_network.capab), msg+offset,
-			nas->elements[4].ms_network.len);
-
-		break;
+		/*Loop through till end of nas messages*/
+		while(offset < nas_msg_len)
+		{
+		    switch((NasIeType_e)msg[offset])
+		    {
+			case NAS_IE_TYPE_MS_NETWORK_CAPABILITY:
+			{
+		            log_msg(LOG_INFO, "MS NW Capability IE \n");
+			    index++;
+			    nas->elements[index].msgType = NAS_IE_TYPE_MS_NETWORK_CAPABILITY;
+			    nas->elements[index].pduElement.ms_network.pres = true;
+			    nas->elements[index].pduElement.ms_network.element_id 
+						= msg[offset];
+			    ++offset;
+			    nas->elements[index].pduElement.ms_network.len 
+						= msg[offset];
+			    ++offset;
+			    memcpy(
+			     (nas->elements[index].pduElement.ms_network.capab), 
+			      msg+offset, 
+			      nas->elements[index].pduElement.ms_network.len);
+			    offset += 
+			      nas->elements[index].pduElement.ms_network.len;
+			    break;
+			}
+			default:
+			{
+			    log_msg(LOG_ERROR, "UnHandled NAS IE type- 0x%x\n", msg[offset]);
+			    offset++;
```

Comment:
Dont we need to skip the length of the IE ? if we do offset++ then we might lose the track.

Reply:
We don't know which IE it is. So, We wouldn't know if it has a length value or not. So not skipping. Its an error case scenario.

Reply:
Which spec has this IE type & encoding defined ?